app/models/manageiq/providers/openstack/inventory/parser/infra_manager.rb

```diff
@@ -0,0 +1,366 @@
+      }
+    end
+  end
+
+  def server_address(server, key)
+    # TODO(lsmola) Nova is missing information which address is primary now,
+    # so just taking first. We need to figure out how to identify it if
+    # there are multiple.
+    server&.addresses&.fetch_path('ctlplane', 0, key)
+  end
+
+  def get_purpose(instance_uuid)
+    collector.stack_resources_by_id[instance_uuid]&.dig('resource_name')
+  end
+
+  def identify_product(instance_uuid)
+    purpose = get_purpose(instance_uuid)
+    return nil unless purpose
+
+    if purpose == 'NovaCompute'
+      'rhel (Nova Compute hypervisor)'
+    else
+      "rhel (No hypervisor, Host Type is #{purpose})"
+    end
+  end
+
+  def identify_host_name(instance_uuid, uid)
+    purpose = get_purpose(instance_uuid)
+    return uid unless purpose
+
+    "#{uid} (#{purpose})"
+  end
+
+  def identify_primary_mac_address(host)
+    server = collector.servers_by_id[host.instance_uuid]
+    server_address(server, 'OS-EXT-IPS-MAC:mac_addr')
+  end
+
+  def identify_primary_ip_address(host)
+    server = collector.servers_by_id[host.instance_uuid]
+    server_address(server, 'addr')
+  end
+
+  def identify_ipmi_address(host)
+    host.driver_info["ipmi_address"]
+  end
+
+  def identify_hypervisor_hostname(host)
+    collector.servers_by_id[host.instance_uuid]&.name
+  end
+
+  def lookup_power_state(power_state_input)
+    case power_state_input
+    when "power on"               then "on"
+    when "power off", "rebooting" then "off"
+    else                               "unknown"
+    end
+  end
+
+  def lookup_connection_state(power_state_input)
+    case power_state_input
+    when "power on"               then "connected"
+    when "power off", "rebooting" then "disconnected"
+    else                               "disconnected"
+    end
+  end
+
+  def clusters_and_host_mapping
+    clusters = []
+    cluster_host_mapping = {}
+    orchestration_stacks = @data_index.fetch_path(:orchestration_stacks)
+    orchestration_stacks&.each_value do |stack|
+      parent = orchestration_stacks[stack[:parent]&.stringified_reference]
+      next unless parent
+
+      nova_server = stack[:resources].detect { |r| collector.stack_server_resource_types.include?(r[:resource_category]) }
+      next unless nova_server
+
+      cluster_host_mapping[nova_server[:physical_resource]] = parent[:ems_ref]
+      clusters << {:name => parent[:name], :uid => parent[:ems_ref]}
+    end
+    return clusters.uniq, cluster_host_mapping
+  end
+
+  def parse_cluster(cluster)
+    name = cluster[:name]
+    uid = cluster[:uid]
+
+    new_result = {
+      :ems_ref => uid,
+      :uid_ems => uid,
+      :name    => name
+    }
+
+    persister.clusters.build(new_result)
+
+    return uid, new_result
+  end
+
+  def set_relationship_on_hosts(hosts, cluster_host_mapping)
+    hosts.each do |host|
+      host.ems_cluster = persister.clusters.lazy_find(cluster_host_mapping[host[:uid_ems]])
```

Comment:
This I'd also like to move to the collector, something like `cluster_by_host_uuid` so that it can be set by the host parser which would allow us to do targeted refresh in the future

Reply:
Also fixed, we set the link to the cluster in the host parser now